### ISSUE TITLE
Implement *-discarded actions

### ISSUE THREAD

**Issue by vladox** [kdeldycke/mail-deduplicate#146]
#### Preliminary checks

* [x] I am running the latest version
* [x] I have [read the Code of Conduct](https://github.com/kdeldycke/mail-deduplicate/blob/develop/.github/code-of-conduct.md)
* [x] I have checked there is not other [Issues](https://github.com/kdeldycke/mail-deduplicate/issues) or [Pull Requests](https://github.com/kdeldycke/mail-deduplicate/pulls) covering the same topic to open

#### Describe the bug

When running `mdedup -a delete-discarded -s select-one ./.MyMailDirFolder` I get the follwoing result:

**NotImplementedError: delete-discarded action not implemented yet.**

#### To reproduce

1. The full `mdedup -a delete-discarded -s select-one ./.MyMailDirFolder` CLI invocation you used.

   ```shell-session
   $ mdedup -a delete-discarded -s select-one ./.MyMailDirFolder
   ```

1. The data set leading to the bug.
   Try to produce here the minimal subset of mails leading to the bug, and add copies of those mails (eventually censored).
   This effort will help maintainers add this particular edge-case to the set of unittests to prevent future regressions.
   You can reduce down the issue to a particular deduplicate subset by using the `--hash-only` parameter.

#### Expected behavior

Duplicated emails being deleted

Is this an expected behaviour?

**Comment by leggewie** [comment]
same is true for move-discarded

**Comment by andy-shev** [comment]
So the core functionality of the tool is not implemented... Can we prioritize this bug (which actually makes tool useless) to be fixed sooner than later?

**Comment by kdeldycke** [comment]
@vincentbernat is right. All `*-discarded` actions are not implemented yet.

In the mean time, you can indeed emulate the effect of `-a delete-discarded -s select-one` with `-a delete-selected -s select-all-but-one`. I.e. by reversing the logic.

The thing is all `*-discarded` actions are syntactic sugar. They should implement the reverse of `*-selected` actions at:
https://github.com/kdeldycke/mail-deduplicate/blob/f423b414c400d21e338edf09bf904d39697fc7b8/mail_deduplicate/action.py#L82-L91

Same for strategies by the way. They all have their own aliases to allow users to better map their own worldview to the operational logic. See:
https://github.com/kdeldycke/mail-deduplicate/blob/f423b414c400d21e338edf09bf904d39697fc7b8/mail_deduplicate/strategy.py#L206-L224

The problem is at the selection process step, right before we perform the action:
https://github.com/kdeldycke/mail-deduplicate/blob/f423b414c400d21e338edf09bf904d39697fc7b8/mail_deduplicate/deduplicate.py#L277

Applying the strategy return the list of selected mails, not the one that were discarded. See how each implemented strategies only returns the subset of the duplicate pool:
https://github.com/kdeldycke/mail-deduplicate/blob/f423b414c400d21e338edf09bf904d39697fc7b8/mail_deduplicate/strategy.py#L37-L39
https://github.com/kdeldycke/mail-deduplicate/blob/f423b414c400d21e338edf09bf904d39697fc7b8/mail_deduplicate/strategy.py#L52-L56
and so on...

By the time we need to perform the action, we only have a subset of the initial duplicate pool. We do not have the list of those that were discarded.

This is the limit of the code architecture inherited from the initial dumb script I wrote 10 years ago. It was an early optimization to reduce the memory footprint. Given that context, it will be hard to easily implement the action with the current code structure.

I propose to first tackle #87, i.e. keep a cache of canonical hashes used to ID each mail. That way we'll be in a position to only deal with sets of hashes in our selection/action phases instead of parsed mail objects. This will bring much cleaner and flexible code to implement the missing actions.

**Comment by andy-shev** [comment]
@kdeldycke 
The problem is that single mails are got removed by your proposal. So, it's not equivalent substitution, please revert "bug" tag and prioritize the fix.

**Comment by nutria007** [comment]
I've managed to do the inteded action by making it skip unique emails:

**mail_deduplicate/mail_deduplicate/deduplicate.py**

```
            # Unique mails are always selected. No need to mobilize the whole
            # DuplicateSet machinery.
            if mail_count == 1:
                self.stats["mail_unique"] += 1
                self.stats["set_single"] += 1
                if self.conf.action == "delete-selected":
                    logger.debug("Skipped deletion of unique mail.")
                    self.stats["mail_skipped"] += 1
                else:
                    logger.debug("Add unique message to selection.")
                    self.stats["mail_selected"] += 1
                    candidates = mail_set
            # We need to resort to a selection strategy to discriminate mails
            # within the set.
            else:
                duplicates = DuplicateSet(hash_key, mail_set, self.conf)
                candidates = duplicates.select_candidates()
                # Merge duplicate set's stats to global stats.
                self.stats += duplicates.stats
```

**Comment by andy-shev** [comment]
> I've managed to do the inteded action by making it skip unique emails:
> 
> **mail_deduplicate/mail_deduplicate/deduplicate.py Lines 404 to 422**

Thanks, it;s exact place where I'm coding right now for myself, but on top I introduced a new option.

**Comment by dschrempf** [comment]
Any progress on this? `select-newer` also doesn't work as expected. Single mails without duplicates are still selected.
```
│ Mails      │   Metric │ Description                                                  │
╞════════════╪══════════╪══════════════════════════════════════════════════════════════╡
│ Found      │     1060 │ Total number of mails encountered from all mail sources.     │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Rejected   │        0 │ Number of mails rejected individually because they were      │
│            │          │ unparseable or did not have enough metadata to compute       │
│            │          │ hashes.                                                      │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Retained   │     1060 │ Number of valid mails parsed and retained for deduplication. │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Hashes     │     1060 │ Number of unique hashes.                                     │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Unique     │     1060 │ Number of unique mails (which where automatically added to   │
│            │          │ selection).                                                  │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Duplicates │        0 │ Number of duplicate mails (sum of mails in all duplicate     │
│            │          │ sets with at least 2 mails).                                 │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Skipped    │        0 │ Number of mails ignored in the selection phase because the   │
│            │          │ whole set they belong to was skipped.                        │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Discarded  │        0 │ Number of mails discarded from the final selection.          │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Selected   │     1060 │ Number of mails kept in the final selection on which the     │
│            │          │ action will be performed.                                    │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Copied     │        0 │ Number of mails copied from their original mailbox to        │
│            │          │ another.                                                     │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Moved      │        0 │ Number of mails moved from their original mailbox to         │
│            │          │ another.                                                     │
├────────────┼──────────┼──────────────────────────────────────────────────────────────┤
│ Deleted    │     1060 │ Number of mails deleted from their mailbox in-place.         │
```
Command line was
```
 mdedup -s select-newer -a delete-selected -n -t ctime maildir
```

**Comment by leggewie** [comment]
@dschrempf Can you please open a new ticket regarding select-newer misbehaving?

**Comment by djwf** [comment]
FYI: just added an issue and PR regarding the option to only act on duplicates (#203 and #204, respectively)

**Comment by kdeldycke** [comment]
The missing `*-discarded` actions have been implemented in #290 and are now available upstream.
